### PR TITLE
Added option to copy code snippet to clipboard

### LINKED ISSUE #1259: Consider adding a copy button
In the previous UX studies and [user feedback](https://github.com/flutter/website/issues/2397) on Github, opportunistic learners appreciate the copy button in code snippets. 

Since we're adding embededd DartPad to the API doc and serve it as code samples, we can consider adding a similar copy button.

Reference below are from current flutter.dev and api.flutter.dev 
image

image

### PR DESCRIPTION
**Pull request by jawahars16**
Added copy option to the code editor, to copy code snippet. Code snippet from the active tab will be copied to clipboard, in case of multiple tabs. Copy code button appear only on hovering the code editor.

image

Fixed #1259

### CODE REVIEW

File: lib/experimental/new_embed.dart [hunk 1]
```diff
@@ -607,6 +612,43 @@ major browsers, such as Firefox, Edge (dev channel), or Chrome.
...
+    String activeSource;
+    String activeTab = "editor";
+
+    try {
+      // TODO: Remove this try-catch block
+      // by fixing the selectedTab to return first tab, if selected attr not available.
+      activeTab = tabController.selectedTab.name;
+    } catch (_) {}
+
+    switch (activeTab) {
```
Comment (RedBrogdon) [inline lib/experimental/new_embed.dart]: I believe this is missing two other possibilities: the test method and the solution.

Have a look at this URL in your build of the site:

`/embed-flutter.html?id=2234a5ccada200eb1e018b12fa95d57d`

There is an additional solution tab, and if you use the menu in the top right, you can enable the test tab as well.

File: lib/experimental/new_embed.dart [hunk 2]
```diff
@@ -607,6 +612,43 @@ major browsers, such as Firefox, Edge (dev channel), or Chrome.
...
+  }
+
+  String _getActiveSourceCode() {
+    String activeSource;
+    String activeTab = "editor";
+
+    try {
+      // TODO: Remove this try-catch block
+      // by fixing the selectedTab to return first tab, if selected attr not available.
+      activeTab = tabController.selectedTab.name;
```
[the remark this answers is not in the document]
Comment (jawahars16) [reply lib/experimental/new_embed.dart]: Updated base `TabController` and removed the `try catch` block. Also added `selected` attribute to first tab.

File: lib/experimental/new_embed.dart [hunk 3]
```diff
@@ -607,6 +612,43 @@ major browsers, such as Firefox, Edge (dev channel), or Chrome.
...
+    textElement.value = _getActiveSourceCode();
+    document.body.append(textElement);
+    textElement.select();
+    document.execCommand('copy');
+    textElement.remove();
+  }
+
+  String _getActiveSourceCode() {
+    String activeSource;
+    String activeTab = "editor";
```
[the remark this answers is not in the document]
Comment (jawahars16) [reply lib/experimental/new_embed.dart]: Renamed variable to `activeTabName`

File: web/experimental/styles.scss [hunk 4]
```diff
@@ -185,11 +185,35 @@ body {
 #editor-and-console-container {
   display: flex;
   flex-direction: column;
+  position: relative;
+}
+
+#copy-code {
```
[the remark this answers is not in the document]
Comment (jawahars16) [reply web/experimental/styles.scss]: Updated button color to match the Run button.

File: lib/experimental/new_embed.dart (hunk 1 above)
Comment (jawahars16) [reply lib/experimental/new_embed.dart]: Considered solution and test tab also.

File: lib/elements/elements.dart [hunk 5]
```diff
@@ -560,7 +560,7 @@ class TabController {
   }
 
   TabElement get selectedTab =>
-      tabs.firstWhere((tab) => tab.hasAttr('selected'));
+      tabs.firstWhere((tab) => tab.hasAttr('selected'), orElse: () => null);
```
Comment (johnpryan) [inline lib/elements/elements.dart]: Can you help me understand this change? Were you seeing an exception thrown here? 

There *should* always be a selected tab. I'm worried this change could be masking over a timing issue.

File: lib/experimental/new_embed.dart [hunk 6]
```diff
@@ -623,15 +623,10 @@ major browsers, such as Firefox, Edge (dev channel), or Chrome.
 
   String _getActiveSourceCode() {
     String activeSource;
-    String activeTab = "editor";
+    String activeTabName = tabController.selectedTab == null ? 
+    'editor' : tabController.selectedTab.name;
```
Comment (RedBrogdon) [inline lib/experimental/new_embed.dart]: nit: I don't think these lines are formatted correctly. Try using `dartfmt` on them.

File: lib/elements/elements.dart (hunk 5 above)
Comment (jawahars16) [reply lib/elements/elements.dart]: Earlier the tabs only has `aria-selected` attribute and not `selected`. So it was throwing `Bad state` error for the first time load. I realised after adding `selected` attribute to the default tab, this null check is not required anymore. 😉 Updated PR.

File: lib/experimental/new_embed.dart (hunk 6 above)
Comment (jawahars16) [reply lib/experimental/new_embed.dart]: Formatted code.

**Comment by RedBrogdon** [comment]
Merged. This will go out in our next deployment, probably in a few days.

Thanks again for submitting a great feature!

**Pull request merged**